proposals/20230605-rules-adoption-management-maturity-framework.md

```diff
@@ -0,0 +1,222 @@
+
+Rules maturity will adopt the [status](https://github.com/falcosecurity/evolution/blob/main/REPOSITORIES.md#status) levels used to indicate the maturity of each repository within The Falco Project, that is, "Stable", "Incubating", "Sandbox", "Deprecated":
+
+- **Stable** (enabled by default) indicates that the rule has been thoroughly evaluated by experts in the field who have hands-on production experience. These experts have determined that the rule addresses a highly relevant threat and recommend that adopters adjust the rule to their specific environment. "Stable" rules embody the utmost best practices and demonstrate optimal robustness in relation to Falco's current capabilities.      
+- **Incubating** (disabled by default) indicates that rules have been identified by experts as catering to more specific use cases that may or may not be relevant for each adopter. However, these rules address relevant threats, provide some level of robustness guarantee, and adhere to best practices in rule writing.    
+- **Sandbox** (disabled by default), indicates that the rule is in an experimental stage. The potential for broader usefulness and relevance of "Sandbox" rules is currently being assessed. The rule can serve as an inspiration and adheres to the minimum acceptance criteria for rules.
+- **Deprecated** (disabled by default), indicates that after re-assessing the rule, it was deemed less applicable to the broader community, and each adopter needs to determine their relevance on their own. These rules are kept as examples, but they are no longer actively supported or tuned by The Falco Project.
+
+> Falco introduces a new key called "maturity" to each rule with levels including "Stable", "Incubating", "Sandbox" and "Deprecated" to guide the adoption process. This key reflects the robustness and relevance of each predefined rule in the [falcosecurity/rules](https://github.com/falcosecurity/rules/blob/main/rules/) repository. Each rule must go through the "Incubating" state before advancing to "Stable" as a minimum requirement. Only "Stable" rules will be enabled by default. 
+
+The maturity level of the rules, however, does not directly reflect their potential for generating noise in the adopters' environment. This is due to the unique and constantly changing nature of each environment, especially in cloud environments, making it challenging to accurately predict the impact of rules.
+
+Newcomers to Falco will be encouraged to start by configuring their setup with introductory rules labeled as "Falco's default rules" (maturity: "Stable"). These rules will be based on syscall and container events and will live in the established [falco_rules.yaml](https://github.com/falcosecurity/rules/blob/main/rules/falco_rules.yaml). 
```

Comment:
How we could / would want to distinguish the maturity level (e.g. `stable`) from the importance / probability that those rules are as agnostic as possible and likely useful for a large number of environments?
For example we can have stable Falco rules, and stable but environment-specific rules.

Is it enough to split Falco rules from application rules?

Reply:
Happy to discuss, for me from my perspective "Stable" shall not indicate stable from a software engineering perspective, but rather tags rules that we recommend any adopter regardless of environment or industry shall adopt. The "Stable" rules aim to detect indicators of compromise of top cyber threats relevant to any environment.

Re your second point I actually would love to understand what "application" means for everyone, because I am confused by for example the existing application rules https://github.com/falcosecurity/rules/blob/main/rules/application_rules.yaml as they are all based on network related system calls and the "application" here comes from the port definitions. An alternative way of looking at application could be to really only focus on higher up in the stack such as auth bypasses ...

Perhaps we should clearly define "system" and "application" level detections for Falco? Another example are detections around Kube audit logs, they are more "control plane" ... in summary clear definitions we all agree upon would be amazing. @maxgio92 would you have a suggestion for those definitions?

Reply:
Thanks @incertum, I got the meaning of stable here, so it comprises also the relevance.
What creates confusion a bit is the fact that I'd expect the stability levels (`stable`, `incubating`, `sandbox`) would be transient - e.g. incubating rules might become stable, sandbox ones incubating.
If I undestand correctly, instead, these levels are not transient - e.g. incubating rules whether correct or not, are supposed to be relevant only for a subset of environment. Am I wrong?

I agree with the fact that we should clarify the distinction between [application](https://github.com/falcosecurity/rules/blob/main/rules/application_rules.yaml) and [system / core](https://github.com/falcosecurity/rules/blob/main/rules/falco_rules.yaml) rules.
From my understanding, application ones are what are not i.e. stable, in the sense of not being rules that should detect threats relevant to any environment, but instead, threats relevant to specific environments, and specifically related to optional user space applications, in some way.
I don't think, as far as I undestand, application refers to the stack level, in the OSI sense.

> Then, additional data sources, through plugins, can relate to additional application rules (like Kubernetes Audit rules with related plugin), but application rules might not necessary need plugins.

In any case I think we should clarify the meaning of application here. And I'd like to map this distinction (system/core and application) to the stability levels (stable, incubating, sandbox, deprecated).
If the above is correct, I think it could make sense and make the vocabulary more consistent. What do you think?

Reply:
@maxgio92 I agree, using the status levels felt appealing, but maybe they not quite right yet. Good news is I think we appear to align on what we are looking for, just need to find a good leveling that is expressive, self-explanatory and creates the least confusion - with that do you have ideas? 🙃 

I like your thinking wrt to defining "application" a lot actually! So in summary what would you suggest we change it to? I think we are close, just need  maybe 2-3 more rounds of editing.

Reply:
I propose to mimic the current repository [scope](https://github.com/falcosecurity/evolution/blob/main/REPOSITORIES.md#scope) **and** [status](https://github.com/falcosecurity/evolution/blob/main/REPOSITORIES.md#status) parameters.

Scopes: _core_ (e.g. [core rules](https://github.com/falcosecurity/rules/blob/main/rules/falco_rules.yaml)) and _ecosystem_ (e.g. [application rules](https://github.com/falcosecurity/rules/blob/main/rules/application_rules.yaml) and matching the *application* meaning which applies to application rules), that apply to rule set and refers to the relevance of the rules for the community use cases.

And for each rule set, orthogonal to the scope, the status maturity level, that represents the maturity stage of the rules (_stable_, _incubating_, _sandbox_, _deprecated_), of which incubating and sandbox ones likely try to converge to the stable level.

I'd like to have @leogr's feedback too, who worked on the parallel parameters for the repositories, and sure that can add important feedback :-)

Reply:
This is great, in summary 2 new keys `maturity` and `scope`. Would try to adjust the wordings in a way that still conveys that transitioning to "stable" maturity has to provide some broader value than string matching on fields in a particular context that could be easily circumvented. Would this resonate with all of you?

Perhaps the system vs application discussion is another discussion we should continue, but it does not have to block this framework. Could see a solution where we just have all rules that are *not* plugin rules in the same file for easier maintenance. Distinctions shall be based on the new keys and possible future extensions of the tags key?

Yes @leogr plz share your perspective, ty!

Reply:
I have made some adjustments, PTAL, thanks.

Reply:
Thank you both for the discussion.

I appreciate the idea of trying to make it clearer to users and contributors the purpose of each file we distribute. I took a look at the `maturity` and `scope` paragraphs and the difference wasn't completely clear to me. I would propose we start with one level, `maturity` (stable, incubating, sandbox), that we can also use to distribute the rules (@leogr and @incertum had a great idea about the distribution part). This would allow us to move forward, be clearer so people don't have to figure out the differences between the two parts, make the repo easier to maintain and contribute to. If we understand that we need another level we can add it at a future iteration.

Reply:
Considering all feedback I have pushed a new version. I like the callout of  keeping things more simple in the beginning.